### PR TITLE
[MJAVADOC-642] do not try to read from links in offline mode.

### PR DESCRIPTION
**Pull request by bmarwell**
- output error stack traces in debug mode.

Pulled out a trimmed-lowercase-link variable for readability and added a null check which oddly occured after creating this PR.
Please note that not even the non-integration-tests can be run behind a proxy.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MJAVADOC) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MJAVADOC-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MJAVADOC-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify -Prun-its` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by rfscholte** [comment]
Please don't split the check for package-list and element-list. If there's no element-list you shouldn't log, as the package-list could exist. Only if oth are missing, there's a reason to log.

**Review by rfscholte — CHANGES_REQUESTED** [review]
combine package-list/element-list message to prevent unnecesary noise in logging

**Comment by bmarwell** [comment]
> combine package-list/element-list message to prevent unnecesary noise in logging

Will do (revert). I thought it might have been useful for debugging (-X), but this is probably out of scope for this PR.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 1]
```diff
@@ -6677,6 +6682,11 @@ protected boolean isValidJavadocLink( String link, boolean detecting )
                 elementListUri = new File( dir, ELEMENT_LIST ).toURI();
             }
 
+            if ( isOffline && !trimmedLowercaseLink.startsWith( "file:" ) )
+            {
+              // checking of both elementlists and package lists skipped.
+              return false;
```
Comment (rfscholte) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6688]: `localhost` and `127.0.0.1` are considered safe too. Makes we wonder if there's a way to use Mavens own Wagon instance that already respects the configuration of settings.xml

Comment (bmarwell) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6732]: @rfscholte do you have any explicit suggestion how to resolve this? Technically "offline" could also mean "no network", where 127.0.0.1 is a local network interface, too. But if there is a better solution, I would be willing to implement it.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 2]
```diff
@@ -1457,21 +1457,21 @@
...
+     * </pre>
      * @since 3.3.0
-     */
-    @Parameter
-    private String[] addStylesheets;
-    
+     */
+    @Parameter
+    private String[] addStylesheets;
+
```
Comment (bmarwell) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:1474]: Mixed windows linebreaks converted to unix by IDE.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 3]
```diff
@@ -3119,34 +3119,34 @@ private void addAddStyleSheets( List<String> arguments ) throws MavenReportExcep
...
+                                                        + "than stylesheetfile: " + stylesheetfile.get().getName() );
+                }
+            }
+            
+            return Optional.of( addstylesheetfile );
+        }
+
+        throw new MavenReportException( "additional stylesheet file does not exist: " 
+                                            + addstylesheetfile.getAbsolutePath() );
+    }
```
Comment (bmarwell) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:3149]: Mixed windows linebreaks converted to unix by IDE.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 4]
```diff
@@ -5643,7 +5643,7 @@ private void addStandardDocletOptions( File javadocOutputDirectory,
                               JavadocUtil.quotedPathArgument( stylesheetfile.get().getAbsolutePath() ) );
         }
         
-        addAddStyleSheets( arguments ); 
+        addAddStyleSheets( arguments );
```
Comment (bmarwell) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:5646]: Mixed windows linebreaks converted to unix by IDE.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 1 above)
Comment (bmarwell) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6732]: Well, a quick interim solution could be sth like this:

```diff
Index: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java b/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java
--- a/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java	(revision 92957646c0be81474fcbf1e84032880f2e2b350e)
+++ b/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java	(date 1621452882333)
@@ -4368,7 +4368,7 @@
                 continue;
             }
 
-            if ( isOffline && !link.startsWith( "file:" ) )
+            if ( isOffline && isOnlineLink( link ))
             {
                 continue;
             }
@@ -4382,6 +4382,19 @@
         }
     }
 
+    private boolean isOnlineLink( String link ) {
+        if  ( link.startsWith( "file:" ) ) {
+            return true;
+        }
+
+        if ( link.startsWith( "http" ) &&
+            (link.contains( "//localhost" ) || link.contains( "//127.0.0." ) ) ) {
+            return false;
+        }
+
+        return true;
+    }
+
     /**
      * Coppy all resources to the output directory
      *
@@ -6726,7 +6739,7 @@
                 elementListUri = new File( dir, ELEMENT_LIST ).toURI();
             }
 
-            if ( isOffline && !trimmedLowercaseLink.startsWith( "file:" ) )
+            if ( isOffline && isOnlineLink( trimmedLowercaseLink ))
             {
               // checking of both elementlists and package lists skipped.
               return false;
```

But I fear such provisional changes, as nothings holds as forever as provisionally solutions.

**Comment by rfscholte** [comment]
When this plugin wants to use the settings' `offline`, it should behave exactly the same as Maven uses this property, which means Wagon should be used. This will completely replace all httpclient code. I can make a new ticket which will superseed this one if you want.

**Comment by bmarwell** [comment]
Sure, go ahead. There's more to change my introducing wagon which would be out of scope in this issue.

**Comment by cstamas** [comment]
my 5 cents: a big NO to use Wagon (maven2) stuff in any maven plugin, not is just an overkill, but wrong IMO as well...
Btw "it should behave exactly the same as Maven uses this property, which means Wagon should be used", does it means that the logic defining Maven behavior re offline is not in Maven but in Wagon?

**Comment by rfscholte** [comment]
I should have said: the same implementation handling URL requests. Based of https://maven.apache.org/resolver/index.html that could still be Wagon. Based on the image we should be using the same "transport", right?

**Comment by bmarwell** [comment]
> I should have said: the same implementation handling URL requests. Based of https://maven.apache.org/resolver/index.html that could still be Wagon. Based on the image we should be using the same "transport", right?

This PR is about LESS artefacts to transport. Why change the transport in such a PR?

**Comment by rfscholte** [comment]
The transport layer is the central location to understand offline.

**Comment by jira-importer** [comment]
Resolve #970

**Pull request closed without merging**